extern/storage-sealing/states_replica_update.go

```diff
@@ -0,0 +1,184 @@
+	params := &miner.ProveReplicaUpdatesParams{
+		Updates: []miner.ReplicaUpdate{
+			{
+				SectorID:           sector.SectorNumber,
+				Deadline:           sl.Deadline,
+				Partition:          sl.Partition,
+				NewSealedSectorCID: *sector.UpdateSealed,
+				Deals:              sector.dealIDs(),
+				UpdateProofType:    updateProof,
+				ReplicaProof:       sector.ReplicaUpdateProof,
+			},
+		},
+	}
+	if err := params.MarshalCBOR(enc); err != nil {
+		log.Errorf("failed to serialize update replica params: %w", err)
+		return ctx.Send(SectorSubmitReplicaUpdateFailed{})
+	}
+
+	// XXX fees need to be better handled
+	mi, err := m.Api.StateMinerInfo(ctx.Context(), m.maddr, tok)
+	if err != nil {
+		log.Errorf("handleSubmitReplicaUpdate: api error, not proceeding: %+v", err)
+		return nil
+	}
+
+	from, _, err := m.addrSel(ctx.Context(), mi, api.CommitAddr, big.Zero(), big.Zero())
+	if err != nil {
+		log.Errorf("no good address to send replica update message from: %+v", err)
+		return ctx.Send(SectorSubmitReplicaUpdateFailed{})
+	}
+	// XXX fees need to be better handled
+	mcid, err := m.Api.SendMsg(ctx.Context(), from, m.maddr, miner.Methods.ProveReplicaUpdates, big.Zero(), big.Int(m.feeCfg.MaxCommitGasFee), enc.Bytes())
+	if err != nil {
+		log.Errorf("handleSubmitReplicaUpdate: error sending message: %+v", err)
+		return ctx.Send(SectorSubmitReplicaUpdateFailed{})
+	}
+
+	return ctx.Send(SectorReplicaUpdateSubmitted{Message: mcid})
+}
+
+func (m *Sealing) handleReplicaUpdateWait(ctx statemachine.Context, sector SectorInfo) error {
+	if sector.ReplicaUpdateMessage == nil {
+		panic("handling error states and events")
+	}
+
+	mw, err := m.Api.StateWaitMsg(ctx.Context(), *sector.ReplicaUpdateMessage)
+	if err != nil {
+		log.Errorf("handleReplicaUpdateWait: failed to wait for message: %+v", err)
+		return ctx.Send(SectorSubmitReplicaUpdateFailed{})
+	}
+
+	switch mw.Receipt.ExitCode {
+	case exitcode.Ok:
+		//expected
+	case exitcode.SysErrInsufficientFunds:
+		fallthrough
+	case exitcode.SysErrOutOfGas:
+		log.Errorf("gas estimator was wrong or out of funds")
+		return ctx.Send(SectorSubmitReplicaUpdateFailed{})
+	default:
+		return ctx.Send(SectorSubmitReplicaUpdateFailed{})
+	}
+	si, err := m.Api.StateSectorGetInfo(ctx.Context(), m.maddr, sector.SectorNumber, mw.TipSetTok)
+	if err != nil {
+		log.Errorf("api err failed to get sector info: %+v", err)
+		return ctx.Send(SectorSubmitReplicaUpdateFailed{})
+	}
+	if si == nil {
+		log.Errorf("api err sector not found")
+		return ctx.Send(SectorSubmitReplicaUpdateFailed{})
+	}
+
+	if !si.SealedCID.Equals(*sector.UpdateSealed) {
+		log.Errorf("mismatch of expected onchain sealed cid after replica update, expected %s got %s", sector.UpdateSealed, si.SealedCID)
+		return ctx.Send(SectorAbortUpgrade{})
+	}
+	return ctx.Send(SectorReplicaUpdateLanded{})
+}
+
+func (m *Sealing) handleFinalizeReplicaUpdate(ctx statemachine.Context, sector SectorInfo) error {
+	return ctx.Send(SectorFinalized{})
```

Comment:
How big is the update cache at this point?

Reply:
For 2KiB within one update cache directory I see: 
```
ls -lh
total 32
-rw-r--r--  1 zenground0  staff    64B Dec 30 13:25 p_aux
-rw-r--r--  1 zenground0  staff   4.0K Dec 30 13:25 sc-02-data-tree-d.dat
-rw-r--r--  1 zenground0  staff    32B Dec 30 13:25 sc-02-data-tree-r-last.dat
-rw-r--r--  1 zenground0  staff   470B Dec 30 13:25 t_aux
```

I'm guessing data-tree-d.dat scales up to be 2x sector size.  Sound right @cryptonemo ?

Reply:
Proper finalization removing sealed data and update cache is the first follow up we should do after the mvp lands.

Reply:
Tree D is 2N-1 nodes, so basically 2x sector size